### PR TITLE
Catch template engine errors

### PR DESCRIPTION
**Pull request by MrPetovan**
Part of #8653

Template engine exceptions are now properly handled. I couldn't help but refactor the whole template engine situation. Renderer methods now have simpler return/parameter types.

### CODE REVIEW

File: src/Core/Renderer.php [hunk 1]
```diff
@@ -66,28 +66,27 @@ class Renderer
...
-	 *
-	 * @return string substituted string
-	 * @throws Exception
+	 * @param string $template
+	 * @param array  $vars
+	 * @return string
+	 * @throws InternalServerErrorException
 	 */
-	public static function replaceMacros($s, array $vars = [])
+	public static function replaceMacros(string $template, array $vars)
```
Comment (annando) [inline src/Core/Renderer.php:76]: Are you sure that the function is always called with a second parameter?

Comment (MrPetovan) [reply src/Core/Renderer.php:76]: Nope, nice catch!

File: src/Core/Renderer.php [hunk 2]
```diff
@@ -66,28 +66,27 @@ class Renderer
...
 
 		$t = self::getTemplateEngine();
 
 		try {
-			$output = $t->replaceMacros($s, $vars);
+			$output = $t->replaceMacros($template, $vars);
 		} catch (Exception $e) {
-			echo "<pre><b>" . __FUNCTION__ . "</b>: " . $e->getMessage() . "</pre>";
-			exit();
+			DI::logger()->critical($e->getMessage(), ['template' => $template, 'vars' => $vars]);
```
Comment (annando) [inline src/Core/Renderer.php:88]: Why "critical" and not "error"? I would define "critical" as some "core meltdown is expected to happen 10 seconds" event.

Comment (MrPetovan) [reply src/Core/Renderer.php:88]: According to the logger functions in the PSR `LoggerInterface`:
```php
    /**
     * Critical conditions.
     *
     * Example: Application component unavailable, unexpected exception.
     *
     * @param string $message
     * @param array  $context
     *
     * @return void
     */
    public function critical($message, array $context = array());

    /**
     * Runtime errors that do not require immediate action but should typically
     * be logged and monitored.
     *
     * @param string $message
     * @param array  $context
     *
     * @return void
     */
    public function error($message, array $context = array());
```
The templating engine failing is an "application component unavailable" and certainly not an "error that does not require immediate action".

What you are describing is either "alert" or "emergency", the two highest log levels even above "critical".

Comment (annando) [reply src/Core/Renderer.php:88]: Okay.

**Comment by annando** [comment]
On a short glance it is looking fine. Had you used it for some hours on your machine with no PHP warnings or errors logged?

**Comment by MrPetovan** [comment]
I've used it on my test instance, yes. I've hit a couple obvious snags that I fixed before I submitted the PR.

The blind spot of this PR is that if the Friendica log isn't enabled, the error messages will vanish in thin air, going against the user-facing message I added about Friendica log. Should I make this part conditional on the current error logging configuration?

**Comment by annando** [comment]
What do you mean with "Should I make this part conditional on the current error logging configuration?"?

**Comment by MrPetovan** [comment]
Should the generic user-facing message depend on the logging configuration? Right now it says to look in the Friendica log, but depending on the admin log settings, it can be empty.

**Comment by annando** [comment]
I'm unsure. It's partly like this "don't show PHP errors" stuff. Probably we could limit this to logged in users - or even only the admin? I wouldn't want to show this to strangers.

**Comment by MrPetovan** [comment]
Only the admin sounds good.

**Comment by MrPetovan** [comment]
Done.

**Pull request merged**